File: data/scripts/feature_engineering.py
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder
from sklearn.impute import SimpleImputer
from sklearn.decomposition import PCA

# Load the cleaned customer data
customer_data = pd.read_csv('../processed/cleaned_customer_data.csv')

# Load transaction data for feature generation
transaction_data = pd.read_csv('../raw/transaction_data.csv')

# Merge customer and transaction data
data = pd.merge(customer_data, transaction_data, on='customer_id', how='left')

# Handling missing values
def handle_missing_values(df):
    # Define strategy for numerical and categorical columns
    num_imputer = SimpleImputer(strategy='mean')
    cat_imputer = SimpleImputer(strategy='most_frequent')

    # Separate numerical and categorical columns
    num_cols = df.select_dtypes(include=[np.number]).columns
    cat_cols = df.select_dtypes(include=['object']).columns

    # Impute missing values
    df[num_cols] = num_imputer.fit_transform(df[num_cols])
    df[cat_cols] = cat_imputer.fit_transform(df[cat_cols])

    return df

data = handle_missing_values(data)

# One-hot encoding for categorical features
def one_hot_encode(df, columns):
    encoder = OneHotEncoder(sparse=False, handle_unknown='ignore')
    encoded_df = pd.DataFrame(encoder.fit_transform(df[columns]))
    encoded_df.columns = encoder.get_feature_names_out(columns)
    df = df.drop(columns, axis=1)
    df = pd.concat([df, encoded_df], axis=1)
    return df

categorical_columns = ['gender', 'country', 'product_type']
data = one_hot_encode(data, categorical_columns)

# Label encoding for target variable
def encode_target_variable(df, target_column):
    encoder = LabelEncoder()
    df[target_column] = encoder.fit_transform(df[target_column])
    return df

data = encode_target_variable(data, 'churn')

# Feature scaling (normalizing numerical features)
def scale_features(df, columns):
    scaler = StandardScaler()
    df[columns] = scaler.fit_transform(df[columns])
    return df

numerical_columns = ['age', 'income', 'transaction_amount', 'transaction_count']
data = scale_features(data, numerical_columns)

# Generate additional features from transaction data
def generate_transaction_features(df):
    # Create features from transaction data
    df['transaction_amount_per_count'] = df['transaction_amount'] / (df['transaction_count'] + 1)
    df['average_transaction'] = df.groupby('customer_id')['transaction_amount'].transform('mean')
    df['transaction_variance'] = df.groupby('customer_id')['transaction_amount'].transform('var')
    
    # Time-based features (with transaction timestamps)
    df['days_since_last_transaction'] = (pd.to_datetime('now') - pd.to_datetime(df['last_transaction_date'])).dt.days
    
    return df

data = generate_transaction_features(data)

# Principal Component Analysis (PCA) for dimensionality reduction
def apply_pca(df, n_components=5):
    pca = PCA(n_components=n_components)
    principal_components = pca.fit_transform(df)
    pca_columns = [f'pca_{i}' for i in range(n_components)]
    pca_df = pd.DataFrame(principal_components, columns=pca_columns)
    return pca_df

pca_columns = ['transaction_amount', 'transaction_count', 'age', 'income']
pca_features = apply_pca(data[pca_columns])
data = pd.concat([data, pca_features], axis=1)

# Time-based features
def generate_time_features(df):
    df['year'] = pd.to_datetime(df['signup_date']).dt.year
    df['month'] = pd.to_datetime(df['signup_date']).dt.month
    df['day_of_week'] = pd.to_datetime(df['signup_date']).dt.weekday
    return df

data = generate_time_features(data)

# Aggregating historical customer data
def aggregate_customer_history(df):
    agg_features = df.groupby('customer_id').agg({
        'transaction_amount': ['mean', 'sum', 'min', 'max'],
        'transaction_count': ['mean', 'sum'],
        'days_since_last_transaction': ['mean', 'min', 'max']
    })
    
    # Flatten the MultiIndex columns
    agg_features.columns = ['_'.join(col).strip() for col in agg_features.columns.values]
    df = pd.merge(df, agg_features, on='customer_id', how='left')
    
    return df

data = aggregate_customer_history(data)

# Interaction terms
def generate_interaction_terms(df):
    df['age_income_interaction'] = df['age'] * df['income']
    df['transaction_amount_income_ratio'] = df['transaction_amount'] / (df['income'] + 1)
    return df

data = generate_interaction_terms(data)

# Drop redundant columns
def drop_redundant_columns(df, columns_to_drop):
    return df.drop(columns_to_drop, axis=1)

columns_to_drop = ['signup_date', 'last_transaction_date']
data = drop_redundant_columns(data, columns_to_drop)

# Save the final engineered dataset
data.to_csv('../processed/engineered_features.csv', index=False)

print("Feature engineering completed and saved.")

File: data/scripts/feature_engineering.R
library(dplyr)
library(tidyr)
library(lubridate)
library(caret)
library(ggplot2)
library(randomForest)
library(MASS)

# Load datasets
customer_data <- read.csv("../raw/customer_data.csv")
transaction_data <- read.csv("../raw/transaction_data.csv")

# Initial data merging (by customer_id)
data <- customer_data %>%
  left_join(transaction_data, by = "customer_id")

# Create new features from existing data
# 1. Time-based features
data$account_age <- as.numeric(difftime(Sys.Date(), as.Date(data$account_creation_date), units = "days"))
data$last_transaction_days <- as.numeric(difftime(Sys.Date(), as.Date(data$last_transaction_date), units = "days"))

# 2. Aggregated transaction features
transaction_summary <- transaction_data %>%
  group_by(customer_id) %>%
  summarise(
    total_transactions = n(),
    total_amount = sum(transaction_amount),
    average_transaction_value = mean(transaction_amount),
    max_transaction_value = max(transaction_amount),
    min_transaction_value = min(transaction_amount),
    std_transaction_value = sd(transaction_amount)
  )

# Merge transaction summary with customer data
data <- data %>%
  left_join(transaction_summary, by = "customer_id")

# 3. Customer interaction features
data$interaction_per_month <- data$total_transactions / (data$account_age / 30)

# Feature engineering for categorical variables
# Convert categorical variables to factors
data$gender <- as.factor(data$gender)
data$churn <- as.factor(data$churn)

# One-hot encoding for 'membership_level'
data <- data %>%
  mutate(
    membership_basic = ifelse(membership_level == "Basic", 1, 0),
    membership_premium = ifelse(membership_level == "Premium", 1, 0),
    membership_gold = ifelse(membership_level == "Gold", 1, 0)
  ) %>%
  select(-membership_level)

# 4. Interaction terms
data$interaction_total_transactions_account_age <- data$total_transactions * data$account_age

# 5. Polynomial features
data$account_age_squared <- data$account_age^2
data$total_transactions_squared <- data$total_transactions^2

# Handling missing values
# Replace missing values with median for numerical features
num_cols <- sapply(data, is.numeric)
data[num_cols] <- apply(data[num_cols], 2, function(x) ifelse(is.na(x), median(x, na.rm = TRUE), x))

# Replace missing values for categorical variables with mode
mode_func <- function(x) {
  ux <- unique(x)
  ux[which.max(tabulate(match(x, ux)))]
}
data$gender[is.na(data$gender)] <- mode_func(data$gender)

# Scaling and normalization
# Normalize numeric features using Min-Max scaling
normalize <- function(x) {
  return((x - min(x)) / (max(x) - min(x)))
}
data[num_cols] <- as.data.frame(lapply(data[num_cols], normalize))

# Feature selection using correlation
corr_matrix <- cor(data[num_cols], use = "complete.obs")
high_corr <- findCorrelation(corr_matrix, cutoff = 0.9)
data <- data %>% select(-all_of(high_corr))

# Random Forest feature importance
set.seed(123)
rf_model <- randomForest(churn ~ ., data = data, importance = TRUE)
importance <- as.data.frame(rf_model$importance)
important_features <- rownames(importance)[order(importance$MeanDecreaseGini, decreasing = TRUE)[1:10]]

# Filter dataset to retain only important features
data <- data %>% select(all_of(important_features), churn)

# Train-test split
set.seed(123)
trainIndex <- createDataPartition(data$churn, p = 0.8, list = FALSE)
train_data <- data[trainIndex, ]
test_data <- data[-trainIndex, ]

# Save processed data
write.csv(train_data, "../processed/train_data.csv", row.names = FALSE)
write.csv(test_data, "../processed/test_data.csv", row.names = FALSE)

# Visualization of feature importance
ggplot(importance, aes(x = reorder(rownames(importance), MeanDecreaseGini), y = MeanDecreaseGini)) +
  geom_bar(stat = "identity") +
  coord_flip() +
  labs(title = "Feature Importance", x = "Features", y = "Mean Decrease in Gini") +
  theme_minimal()

# Principal Component Analysis (PCA) for dimensionality reduction
pca_model <- prcomp(data[, -ncol(data)], center = TRUE, scale. = TRUE)
pca_data <- as.data.frame(pca_model$x)
pca_data$churn <- data$churn

# Explained variance
explained_variance <- summary(pca_model)$importance[2,]
cumulative_variance <- cumsum(explained_variance)
plot(cumulative_variance, type = "b", xlab = "Principal Components", ylab = "Cumulative Explained Variance")

# Save PCA results
write.csv(pca_data, "../processed/pca_data.csv", row.names = FALSE)

# K-means clustering to identify customer segments
set.seed(123)
kmeans_result <- kmeans(data[, -ncol(data)], centers = 3, nstart = 20)
data$cluster <- as.factor(kmeans_result$cluster)

# Visualizing clusters
ggplot(data, aes(x = account_age, y = total_transactions, color = cluster)) +
  geom_point(alpha = 0.7) +
  labs(title = "Customer Segments based on K-Means Clustering") +
  theme_minimal()

# Save final processed data with clusters
write.csv(data, "../processed/final_data_with_clusters.csv", row.names = FALSE)

# Save the R environment for further analysis
save.image(file = "../processed/feature_engineering.RData")